pkg/clustermesh/mcsapi/serviceimport_controller.go

```diff
@@ -0,0 +1,489 @@
+			if meta.SetStatusCondition(&svcExport.Status.Conditions, metav1.Condition{
+				Type:               mcsapiv1alpha1.ServiceExportValid,
+				Status:             metav1.ConditionFalse,
+				LastTransitionTime: metav1.Now(),
+				Reason:             "ServiceType",
+				Message:            "Service type ExternalName is not supported",
+			}) {
+				if err := r.Client.Status().Update(ctx, svcExport); err != nil {
+					return controllerruntime.Fail(err)
+				}
+			}
+			return controllerruntime.Success()
+		}
+	}
+
+	oldestCluster := getOldestRemoteClusterService(svcExportByAllCluster)
+	oldestClusterSvc := svcExportByAllCluster[oldestCluster]
+	mergedPorts := mergePorts(svcExportByAllCluster)
+
+	if svcExport != nil {
+		conflict := false
+		changedCondition := meta.SetStatusCondition(&svcExport.Status.Conditions, metav1.Condition{
+			Type:               mcsapiv1alpha1.ServiceExportValid,
+			Status:             metav1.ConditionTrue,
+			Reason:             mcsapiv1alpha1.ServiceExportValid,
+			Message:            "Service is Valid for export",
+			LastTransitionTime: metav1.Now(),
+		})
+
+		if oldestCluster != r.clusterName {
+			conflictMsg := checkConflictExport(svcExportByAllCluster, r.clusterName, oldestCluster)
+			if conflictMsg == "" {
+				conflictMsg = checkConflictExportPorts(svcExportByAllCluster, mergedPorts, r.clusterName)
+			}
+			if conflictMsg != "" {
+				conflict = true
+				changedCondition = meta.SetStatusCondition(&svcExport.Status.Conditions, metav1.Condition{
+					Type:               mcsapiv1alpha1.ServiceExportConflict,
+					Status:             metav1.ConditionTrue,
+					LastTransitionTime: metav1.Now(),
+					Reason:             mcsapiv1alpha1.ServiceExportConflict,
+					Message:            conflictMsg,
+				}) || changedCondition
+			}
+		}
+		readyStatus := metav1.ConditionFalse
+		if !conflict {
+			readyStatus = metav1.ConditionTrue
+			changedCondition = meta.RemoveStatusCondition(&svcExport.Status.Conditions, mcsapiv1alpha1.ServiceExportConflict) || changedCondition
+		}
+		changedCondition = meta.SetStatusCondition(&svcExport.Status.Conditions, metav1.Condition{
+			Type:               conditionTypeReady,
+			Status:             readyStatus,
+			Reason:             conditionTypeReady,
+			LastTransitionTime: metav1.Now(),
+		}) || changedCondition
+		if changedCondition {
+			if err := r.Client.Status().Update(ctx, svcExport); err != nil {
+				return controllerruntime.Fail(err)
+			}
+		}
+	}
+
+	svcImport.Spec.Ports = mergedPortsToMCSPorts(mergedPorts)
+	svcImport.Spec.Type = oldestClusterSvc.Type
+	svcImport.Spec.SessionAffinity = oldestClusterSvc.SessionAffinity
+	svcImport.Spec.SessionAffinityConfig = oldestClusterSvc.SessionAffinityConfig.DeepCopy()
+	if !svcImportExists {
+		if err := r.Client.Create(ctx, svcImport); err != nil {
+			return controllerruntime.Fail(err)
+		}
+	} else {
+		if err := r.Client.Update(ctx, svcImport); err != nil {
```

Comment:
should this use server-side apply with field managers?

Reply:
IIUC this is when this is useful when there is multiple actors that can update a kubernetes object right? In this specific case we are the sole actor updating ServiceImport ressources so I would say that it is not necessary I think. But correct me if I'm wrong happy to take a look at taking a look at this!

Reply:
Another possibility could be using `controllerutil.CreateOrUpdate`.

Reply:
Makes sense; if there's only one owner, then there's no need to use SSA. Just checking :-)